File: libs/power-log-parser/src/lib/handlers/send-choices-handler.ts
import { ChoiceType } from '@firestone-hs/reference-data';
import type { Helper } from '../helper';
import { Action, Choice, Game, NodeType, SendChoices } from '../models';
import { Regexes } from '../regexes';
import type { ParserState } from '../state/parser-state';

export class SendChoicesHandler {
	static Handle(timestamp: string, data: string, state: ParserState, helper: Helper): void {
		if (state.CurrentGame == null || state.Node == null) {
			return;
		}

		data = data.trim();
		let match = Regexes.SendChoicesChoicetypeRegex.exec(data);
		if (match) {
			const id = match[1];
			const rawType = match[2];
			const type = helper.ParseEnum(ChoiceType, rawType);
			const sendChoices = new SendChoices();
			sendChoices.Choices = [];
			sendChoices.Entity = parseInt(id, 10);
			sendChoices.Type = type;
			sendChoices.TimeStamp = timestamp;
			state.SendChoices = sendChoices;

			if (state.Node.Type === NodeType.Game) {
				(state.Node.Object as Game).AddData(state.SendChoices);
			} else if (state.Node.Type === NodeType.Action) {
				(state.Node.Object as Action).Data.push(state.SendChoices);
			} else {
				throw new Error('Invalid node ' + state.Node.Type + ' -- ' + data);
			}
			return;
		}
		match = Regexes.SendChoicesEntitiesRegex.exec(data);
		if (match) {
			const index = helper.ParseEntity(match[1]);
			const id = helper.ParseEntity(match[2]);
			const choice = new Choice();
			choice.Entity = id;
			choice.Index = index;
			state.SendChoices!.Choices.push(choice);
		}
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Handlers/SendChoicesHandler.cs
#region

using System;
using System.Collections.Generic;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.Meta;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

#endregion

namespace HearthstoneReplays.Parser.Handlers
{
	public class SendChoicesHandler
	{
		private Helper helper;

		public SendChoicesHandler(Helper helper)
        {
			this.helper = helper;
        }

		public void Handle(DateTime timestamp, string data, ParserState state)
        {
            // This happens for the first options when spectating a game that has already started
            if (state.CurrentGame == null || state.Node == null)
            {
                return;
            }

            data = data.Trim();
			var match = Regexes.SendChoicesChoicetypeRegex.Match(data);
			if(match.Success)
			{
				var id = match.Groups[1].Value;
				var rawType = match.Groups[2].Value;
			    var type = helper.ParseEnum<ChoiceType>(rawType);
				state.SendChoices = new SendChoices {Choices = new List<Choice>(), Entity = int.Parse(id), Type = type, TimeStamp = timestamp};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).AddData(state.SendChoices);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(state.SendChoices);
				else
					throw new Exception("Invalid node " + state.Node.Type + " -- " + data);
				return;
			}
			match = Regexes.SendChoicesEntitiesRegex.Match(data);
			if(match.Success)
			{
				var index = helper.ParseEntity(match.Groups[1].Value);
				var id = helper.ParseEntity(match.Groups[2].Value);
				var choice = new Choice {Entity = id, Index = index};
				state.SendChoices.Choices.Add(choice);
			}
		}
	}
}